From fa56b90e3486145f5437bb2de6cc7dd8aa846171 Mon Sep 17 00:00:00 2001
From: Dawei Chien <dawei.chien@mediatek.corp-partner.google.com>
Date: Tue, 8 Feb 2022 14:49:04 +0800
Subject: [PATCH] CHROMIUM: thermal: mediatek: Add thermal zone settings for
 mt8186

Add thermal zone settings for mt8186i SoC.

BUG=b:204226007
TEST=boot KRABBY to read temperature and check thermal throttle

Signed-off-by: Dawei Chien <dawei.chien@mediatek.corp-partner.google.com>
Signed-off-by: Allen-kh Cheng <allen-kh.cheng@mediatek.corp-partner.google.com>
Change-Id: I7d9db90d5a886757ddd21395712897ad46ae2264
Reviewed-on: https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/3446861
Reviewed-by: Hsin-Yi Wang <hsinyi@chromium.org>
Commit-Queue: Hsin-Yi Wang <hsinyi@chromium.org>

[rebase61(tzungbi):
    Squashed:
	FIXUP: CHROMIUM: thermal: mediatek: Add thermal zone settings for mt8186
]
Signed-off-by: Tzung-Bi Shih <tzungbi@chromium.org>
---

Notes:
    478501edf32dc271b2a6be4a266f9945a9af3746
    894174f156d68d56c7f4489fadff79a89d48f1a0

 drivers/thermal/mediatek/soc_temp_lvts.c | 191 ++++++++++++++++++++++-
 1 file changed, 184 insertions(+), 7 deletions(-)

diff --git a/drivers/thermal/mediatek/soc_temp_lvts.c b/drivers/thermal/mediatek/soc_temp_lvts.c
index 81567dc192eb..86c51bc5d8e5 100644
--- a/drivers/thermal/mediatek/soc_temp_lvts.c
+++ b/drivers/thermal/mediatek/soc_temp_lvts.c
@@ -605,10 +605,23 @@ static int prepare_calibration_data(struct lvts_data *lvts_data)
 
 	size = sizeof(buffer);
 	offset = snprintf(buffer, size, "[lvts_cal] num:g_count:g_count_rc ");
+
+	if (offset < 0)
+		return -EINVAL;
+
+	if (offset >= size)
+		return -ENOMEM;
+
 	for (i = 0; i < lvts_data->num_sensor; i++)
 		offset += snprintf(buffer + offset, size - offset, "%d:%d:%d ",
 				   i, cal_data->count_r[i], cal_data->count_rc[i]);
 
+	if (offset < 0)
+		return -EINVAL;
+
+	if (offset >= size)
+		return -ENOMEM;
+
 	buffer[offset] = '\0';
 	dev_info(dev, "%s\n", buffer);
 
@@ -621,8 +634,8 @@ static int get_calibration_data(struct lvts_data *lvts_data)
 	char cell_name[8];
 	struct nvmem_cell *cell;
 	u32 *buf;
-	size_t len;
-	int i, j, index = 0, ret;
+	size_t len = 0;
+	int i, offset, j, index = 0;
 
 	lvts_data->efuse = devm_kcalloc(dev, lvts_data->num_efuse_addr,
 					sizeof(*lvts_data->efuse), GFP_KERNEL);
@@ -630,7 +643,12 @@ static int get_calibration_data(struct lvts_data *lvts_data)
 		return -ENOMEM;
 
 	for (i = 0; i < lvts_data->num_efuse_block; i++) {
-		snprintf(cell_name, sizeof(cell_name), "e_data%d", i + 1);
+		offset = snprintf(cell_name, sizeof(cell_name), "e_data%d", i + 1);
+		if (offset < 0)
+			return -EINVAL;
+
+		if (offset >= sizeof(cell_name))
+			return -ENOMEM;
 		cell = nvmem_cell_get(dev, cell_name);
 		if (IS_ERR(cell)) {
 			dev_err(dev, "Error: Failed to get nvmem cell %s\n", cell_name);
@@ -657,9 +675,7 @@ static int get_calibration_data(struct lvts_data *lvts_data)
 		kfree(buf);
 	}
 
-	ret = prepare_calibration_data(lvts_data);
-
-	return ret;
+	return 0;
 }
 
 static int of_update_lvts_data(struct lvts_data *lvts_data,
@@ -717,6 +733,9 @@ static int of_update_lvts_data(struct lvts_data *lvts_data,
 		return -ENOMEM;
 
 	ret = get_calibration_data(lvts_data);
+	if (ret)
+		lvts_data->cal_data.use_fake_efuse = 1;
+	ret = prepare_calibration_data(lvts_data);
 	if (ret)
 		return ret;
 
@@ -999,10 +1018,23 @@ static int device_read_count_rc_n_v4(struct lvts_data *lvts_data)
 
 	size = sizeof(buffer);
 	offset = snprintf(buffer, size, "[COUNT_RC_NOW] ");
+
+	if (offset < 0)
+		return -EINVAL;
+
+	if (offset >= size)
+		return -ENOMEM;
+
 	for (i = 0; i < lvts_data->num_sensor; i++)
 		offset += snprintf(buffer + offset, size - offset, "%d:%d ",
 				   i, cal_data->count_rc_now[i]);
 
+	if (offset < 0)
+		return -EINVAL;
+
+	if (offset >= size)
+		return -ENOMEM;
+
 	buffer[offset] = '\0';
 	dev_info(dev, "%s\n", buffer);
 
@@ -1451,6 +1483,146 @@ static struct lvts_data mt8195_lvts_data = {
 };
 
 /*==================================================
+ * LVTS MT8186
+ *==================================================
+ */
+
+#define MT8186_NUM_LVTS (ARRAY_SIZE(mt8186_tc_settings))
+#define MT8186_TSBG_DEM_CKSEL_X_TSBG_CHOP_EN (DEVICE_WRITE | RG_TSV2F_CTRL_4 << 8 | 0xFC)
+#define MT8186_TSV2F_CHOP_CKSEL_AND_TSV2F_EN (DEVICE_WRITE | RG_TSV2F_CTRL_2 << 8 | 0xAC)
+#define MT8186_SET_TS_CHOP (DEVICE_WRITE | RG_TSV2F_CTRL_0 << 8 | 0xF7)
+#define MT8186_SET_TSV2F_RSV (DEVICE_WRITE | RG_TSV2F_CTRL_3 << 8 | 0x04)
+
+enum mt8186_lvts_domain {
+	MT8186_AP_DOMAIN,
+	MT8186_NUM_DOMAIN
+};
+
+enum mt8186_lvts_sensor_enum {
+	MT8186_TS1_0,
+	MT8186_TS1_1,
+	MT8186_TS1_2,
+	MT8186_TS1_3,
+	MT8186_TS3_0,
+	MT8186_TS3_1,
+	MT8186_TS3_2,
+	MT8186_TS2_0,
+	MT8186_TS2_1,
+	MT8186_NUM_TS
+};
+
+static void mt8186_device_enable_and_init(struct lvts_data *lvts_data)
+{
+	unsigned int i;
+
+	for (i = 0; i < lvts_data->num_tc; i++) {
+		lvts_write_device(lvts_data, STOP_COUNTING_V4, i);
+		lvts_write_device(lvts_data, SET_COUNTING_WINDOW_20US2_V4, i);
+		lvts_write_device(lvts_data, SET_COUNTING_WINDOW_20US1_V4, i);
+		lvts_write_device(lvts_data, SET_RG_TSFM_LPDLY_V4, i);
+		lvts_write_device(lvts_data, MT8186_TSBG_DEM_CKSEL_X_TSBG_CHOP_EN, i);
+		lvts_write_device(lvts_data, MT8186_TSV2F_CHOP_CKSEL_AND_TSV2F_EN, i);
+		lvts_write_device(lvts_data, SET_TS_RSV_V4, i);
+		lvts_write_device(lvts_data, MT8186_SET_TS_CHOP, i);
+		lvts_write_device(lvts_data, MT8186_SET_TSV2F_RSV, i);
+	}
+
+	lvts_data->counting_window_us = 20;
+}
+
+
+static void mt8186_efuse_to_cal_data(struct lvts_data *lvts_data)
+{
+	struct sensor_cal_data *cal_data = &lvts_data->cal_data;
+
+	cal_data->golden_temp = GET_CAL_DATA_BITMASK(0, 31, 24);
+	cal_data->count_r[MT8186_TS1_0] = GET_CAL_DATA_BITMASK(1, 31, 8);
+	cal_data->count_r[MT8186_TS1_1] = (GET_CAL_DATA_BITMASK(1, 7, 0) << 16) +
+						GET_CAL_DATA_BITMASK(2, 31, 16);
+	cal_data->count_r[MT8186_TS1_2] = (GET_CAL_DATA_BITMASK(2, 15, 0) << 8) +
+						GET_CAL_DATA_BITMASK(3, 31, 24);
+	cal_data->count_r[MT8186_TS1_3] = GET_CAL_DATA_BITMASK(3, 23, 0);
+
+	cal_data->count_r[MT8186_TS2_0] = (GET_CAL_DATA_BITMASK(4, 7, 0) << 16) +
+						GET_CAL_DATA_BITMASK(5, 31, 16);
+	cal_data->count_r[MT8186_TS2_1] = (GET_CAL_DATA_BITMASK(5, 15, 0) << 8) +
+						GET_CAL_DATA_BITMASK(6, 31, 24);
+
+	cal_data->count_r[MT8186_TS3_0] = GET_CAL_DATA_BITMASK(7, 31, 8);
+	cal_data->count_r[MT8186_TS3_1] = (GET_CAL_DATA_BITMASK(7, 7, 0) << 16) +
+						GET_CAL_DATA_BITMASK(8, 31, 16);
+	cal_data->count_r[MT8186_TS3_2] = (GET_CAL_DATA_BITMASK(8, 15, 0) << 8) +
+						GET_CAL_DATA_BITMASK(9, 31, 24);
+
+	cal_data->count_rc[MT8186_TS1_0] = GET_CAL_DATA_BITMASK(4, 31, 8);
+	cal_data->count_rc[MT8186_TS2_0] = GET_CAL_DATA_BITMASK(6, 23, 0);
+	cal_data->count_rc[MT8186_TS3_0] = GET_CAL_DATA_BITMASK(9, 23, 0);
+}
+
+static struct tc_settings mt8186_tc_settings[] = {
+	[0] = {
+		.domain_index = MT8186_AP_DOMAIN,
+		.addr_offset = 0x0,
+		.num_sensor = 4,
+		.sensor_map = {MT8186_TS1_0, MT8186_TS1_1, MT8186_TS1_2, MT8186_TS1_3},
+		.tc_speed = SET_TC_SPEED_IN_US(118, 118, 118, 118),
+		.hw_filter = LVTS_FILTER_2_OF_4,
+		.dominator_sensing_point = SENSING_POINT0,
+		.hw_reboot_trip_point = 105000,
+		.irq_bit = BIT(1),
+	},
+	[1] = {
+		.domain_index = MT8186_AP_DOMAIN,
+		.addr_offset = 0x100,
+		.num_sensor = 2,
+		.sensor_map = {MT8186_TS2_0, MT8186_TS2_1},
+		.tc_speed = SET_TC_SPEED_IN_US(118, 118, 118, 118),
+		.hw_filter = LVTS_FILTER_2_OF_4,
+		.dominator_sensing_point = SENSING_POINT0,
+		.hw_reboot_trip_point = 105000,
+		.irq_bit = BIT(2),
+	},
+	[2] = {
+		.domain_index = MT8186_AP_DOMAIN,
+		.addr_offset = 0x200,
+		.num_sensor = 3,
+		.sensor_map = {MT8186_TS3_0, MT8186_TS3_1, MT8186_TS3_2},
+		.tc_speed = SET_TC_SPEED_IN_US(118, 118, 118, 118),
+		.hw_filter = LVTS_FILTER_2_OF_4,
+		.dominator_sensing_point = SENSING_POINT0,
+		.hw_reboot_trip_point = 105000,
+		.irq_bit = BIT(3),
+	},
+};
+
+static struct lvts_data mt8186_lvts_data = {
+	.num_domain = MT8186_NUM_DOMAIN,
+	.num_tc = MT8186_NUM_LVTS,
+	.tc = mt8186_tc_settings,
+	.num_sensor = MT8186_NUM_TS,
+	.ops = {
+		.efuse_to_cal_data = mt8186_efuse_to_cal_data,
+		.device_enable_and_init = mt8186_device_enable_and_init,
+		.device_enable_auto_rck = device_enable_auto_rck_v4,
+		.device_read_count_rc_n = device_read_count_rc_n_v4,
+		.set_cal_data = set_calibration_data_v4,
+		.init_controller = init_controller_v4,
+	},
+	.feature_bitmap = FEATURE_DEVICE_AUTO_RCK | FEATURE_CK26M_ACTIVE,
+	.num_efuse_addr = 10,
+	.num_efuse_block = 2,
+	.cal_data = {
+		.default_golden_temp = 50,
+		.default_count_r = 19000,
+		.default_count_rc = 5350,
+	},
+	.coeff = {
+		.a = -204650,
+		.b = 204650,
+	},
+};
+
+/*
  *==================================================
  * Support chips
  *==================================================
@@ -1464,11 +1636,15 @@ static const struct of_device_id lvts_of_match[] = {
 		.compatible = "mediatek,mt8195-lvts",
 		.data = (void *)&mt8195_lvts_data,
 	},
+	{
+		.compatible = "mediatek,mt8186-lvts",
+		.data = (void *)&mt8186_lvts_data,
+	},
 	{
 	},
 };
 MODULE_DEVICE_TABLE(of, lvts_of_match);
-/*==================================================*/
+
 static struct platform_driver soc_temp_lvts = {
 	.probe = lvts_probe,
 	.remove = lvts_remove,
@@ -1483,5 +1659,6 @@ static struct platform_driver soc_temp_lvts = {
 module_platform_driver(soc_temp_lvts);
 MODULE_AUTHOR("Yu-Chia Chang <ethan.chang@mediatek.com>");
 MODULE_AUTHOR("Michael Kao <michael.kao@mediatek.com>");
+MODULE_AUTHOR("Dawei Chien <dawei.chien@mediatek.com>");
 MODULE_DESCRIPTION("Mediatek soc temperature driver");
 MODULE_LICENSE("GPL v2");
-- 
2.38.1.584.g0f3c55d4c2-goog

